### ISSUE TITLE
jruby-9.3.2.0 issue with using + on two Concurrent::Set objects

### ISSUE THREAD

**Issue by AndyObtiva** [jruby/jruby#6958]
**Environment Information**
* Operating system:                Mac Catalina
* Ruby implementation:             JRuby 9.3.2.0
* jruby -v: `jruby 9.3.2.0 (2.6.8) 2021-12-01 0b8223f905 Java HotSpot(TM) 64-Bit Server VM 16.0.2+7-67 on 16.0.2+7-67 +jit [darwin-x86_64]`
* uname -a: `Darwin Andys-MacBook-Pro-13.local 19.6.0 Darwin Kernel Version 19.6.0: Thu Oct 29 22:56:45 PDT 2020; root:xnu-6153.141.2.2~1/RELEASE_X86_64 x86_64`
* Installed gem: `concurrent-ruby` version:       1.1.9

I just installed the brand new jruby-9.3.2.0 today and encountered the following issue with `Concurrent::Set` from `concurrent-ruby` gem (not sure if a JRuby issue or gem issue, but it was caused by the latest JRuby for sure)

irb example:
```ruby
jruby-9.3.2.0 :006 > require 'concurrent'
 => true 
jruby-9.3.2.0 :007 > Concurrent::Set.new([:a, :b, :c]) + Concurrent::Set.new([:d, :e, :f])
Traceback (most recent call last):
       16: from org.jruby.dist/org.jruby.ext.set.RubySet.initialize_dup(RubySet.java:308)
       15: from org.jruby.dist/org.jruby.runtime.callsite.SuperCallSite.call(SuperCallSite.java:218)
       14: from org.jruby.dist/org.jruby.ext.set.RubySet$INVOKER$i$1$0$initialize_dup.call(RubySet$INVOKER$i$1$0$initialize_dup.gen)
       13: from org.jruby.dist/org.jruby.ext.set.RubySet.initialize_dup(RubySet.java:308)
       12: from org.jruby.dist/org.jruby.runtime.callsite.SuperCallSite.call(SuperCallSite.java:218)
       11: from org.jruby.dist/org.jruby.ext.set.RubySet$INVOKER$i$1$0$initialize_dup.call(RubySet$INVOKER$i$1$0$initialize_dup.gen)
       10: from org.jruby.dist/org.jruby.ext.set.RubySet.initialize_dup(RubySet.java:308)
        9: from org.jruby.dist/org.jruby.runtime.callsite.SuperCallSite.call(SuperCallSite.java:218)
        8: from org.jruby.dist/org.jruby.ext.set.RubySet$INVOKER$i$1$0$initialize_dup.call(RubySet$INVOKER$i$1$0$initialize_dup.gen)
        7: from org.jruby.dist/org.jruby.ext.set.RubySet.initialize_dup(RubySet.java:308)
        6: from org.jruby.dist/org.jruby.runtime.callsite.SuperCallSite.call(SuperCallSite.java:218)
        5: from org.jruby.dist/org.jruby.ext.set.RubySet$INVOKER$i$1$0$initialize_dup.call(RubySet$INVOKER$i$1$0$initialize_dup.gen)
        4: from org.jruby.dist/org.jruby.ext.set.RubySet.initialize_dup(RubySet.java:308)
        3: from org.jruby.dist/org.jruby.runtime.callsite.SuperCallSite.call(SuperCallSite.java:218)
        2: from org.jruby.dist/org.jruby.ext.set.RubySet$INVOKER$i$1$0$initialize_dup.call(RubySet$INVOKER$i$1$0$initialize_dup.gen)
        1: from org.jruby.dist/org.jruby.ext.set.RubySet.initialize_dup(RubySet.java:307)
Java::JavaLang::StackOverflowError ()
```

It works fine with jruby-9.3.1.0:
```ruby
jruby-9.3.1.0 :001 > require 'concurrent'
 => true 
jruby-9.3.1.0 :002 > Concurrent::Set.new([:a, :b, :c]) + Concurrent::Set.new([:d, :e, :f])
 => #<Concurrent::Set: {:a, :b, :c, :d, :e, :f}> 
```

Help is appreciated.

p.s. Also reported at: https://github.com/ruby-concurrency/concurrent-ruby/issues/928

**Comment by headius** [comment]
@AndyObtiva Confirmed and fixed in #6959. We are just debating when to get the fixed version out to you. If you are ok rolling back to 9.3.2 the only big thing you lose is the Date.parse CVE fix, which can be mitigated by patching the "date" library Ruby code yourself, or ignored if you are not parsing untrusted dates.

We will release this fix in 9.2.21 and 9.3.3 but it is a lot of work to put out a release and we want to wait a bit to see if anything else needs to be fixed.

**Comment by AndyObtiva** [comment]
I can wait for 9.3.3. No worries at all.

**Comment by headius** [comment]
So this actually first shipped in 9.2.20, but apparently nobody using that version attempted to add two set descendants of this sort. I have merged #6959 into jruby-9.2 for release as 9.2.21 and will merge that forward to 9.3 for 9.3.3.

**Comment by Magisus** [comment]
We hit this issue in Puppet Server trying to upgrade to 9.2.20, we'll be awaiting the fix in 9.2.21. Thanks for getting to the bottom of it!
